src/ledctl/ledctl.c

```diff
@@ -549,79 +797,169 @@ static led_status_t _cmdline_ibpi_parse(int argc, char *argv[])
  *
  * @param[in]      argc           number of elements in argv array.
  * @param[in]      argv           command line arguments.
+ * @param[in]      req            structure with request.
  *
  * @return LED_STATUS_SUCCESS if successful, otherwise led_status_t.
  */
-static led_status_t _cmdline_parse_non_root(int argc, char *argv[])
+static led_status_t _cmdline_parse_non_root(int argc, char *argv[], struct request *req)
 {
 	int opt_index, opt = -1;
-	led_status_t status = LED_STATUS_SUCCESS;
+	char *shortopts;
+	struct option *longopts;
+
+	if (argc == 1) {
+		fprintf(stderr, "Program cannot be run without parameters.\n");
+		return LED_STATUS_CMDLINE_ERROR;
+	}
+
+	optind = 1;
+	setup_options(&longopts, &shortopts, possible_params, ARRAY_SIZE(possible_params));
 
+	/* Parse all arguments, invalid ones will be placed on the end of array argv[] */
 	do {
-		opt = getopt_long(argc, argv, shortopt, longopt, &opt_index);
+		opt = getopt_long(argc, argv, shortopts, longopts, &opt_index);
+
 		switch (opt) {
 		case 'v':
-			_ledctl_version();
-			exit(EXIT_SUCCESS);
+			req->chosen_opt = OPT_VERSION;
+			break;
 		case 'h':
-			_ledctl_help();
-			exit(EXIT_SUCCESS);
-		case ':':
-		case '?':
-			return LED_STATUS_CMDLINE_ERROR;
+			req->chosen_opt = OPT_HELP;
+			break;
+#ifdef ENABLE_TEST
+		case 'T':
+			test_params = true;
+			break;
+#endif
+		default:
+			break;
 		}
-	} while (opt >= 0);
 
-	return status;
+	} while (opt != -1);
+
+	free(shortopts);
+	free(longopts);
+
+	return LED_STATUS_SUCCESS;
+}
+
+bool is_ibpi_mode_supported_arg(struct request *req, char *arg)
```

Comment:
This limits functionality of ledctl, however I'm not sure if somebody uses ledctl with /sys/devices/ link.
I would like to get this in separate pull.